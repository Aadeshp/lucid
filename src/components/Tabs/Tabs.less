@Tabs-spacer-height: 3px;
@Tabs-tab-height: 28px;
@Tabs-arrow-width: 8px;

.lucid-Tabs {
	&-bar {
		font-size: @fontSize;
		font-family: @fontFamily;
		list-style-type: none;
		display: flex;
		margin: 0;
		padding: 0;
	}

	&-Tab {
		flex: 1 1 auto;
		cursor: pointer;
		user-select: none;
		margin-bottom: -1px;
		height: @Tabs-tab-height;
		display: flex;
		align-items: center;
		justify-content: center;
		border: 1px solid @color-default-borderColor;
		border-right: 0;
		background-color: @color-white;

		&:last-of-type {
			border-right: 1px solid @color-default-borderColor;
		}

		&:hover {
			background-color: @color-lightGray;
			.Tabs-mixin-tab-arrow-background(@color-lightGray);
		}

		&:active {
			background-image: linear-gradient(#dbdbdb 0, @color-lightGray @Tabs-spacer-height);
		}
	}

	&-Tab-is-progressive-complete {
		color: @color-primary;
	}

	&-Tab-is-disabled {
		color: fade(#333, 50%);
		&:hover {
			background-color: @color-white;
			cursor: default;
			.Tabs-mixin-tab-arrow-background();
		}
		&:active {
			background: @color-white;
		}
	}

	&-Tab-is-progressive {
		position: relative;
		border-width: 1px 0;
		&:first-of-type {
			border-left-width: 1px;
		}
		& + * {
			border-left-width: 0;
		}
		.lucid-Tabs-Tab-arrow{
			z-index: 100;
			height: @Tabs-tab-height;
			width: @Tabs-arrow-width;
			position: absolute;
			bottom: -1px;
			right: -@Tabs-arrow-width;
			.lucid-Tabs-Tab-arrow-line {
				stroke: @color-default-borderColor;
			}
		}
		.Tabs-mixin-tab-line(transparent);
		.Tabs-mixin-tab-arrow-background();
	}

	&-Tab-is-active {
		border-top: none;
		padding-top: 1px;
		color: @color-black;
		background-image: linear-gradient(@color-primary 0, @color-primary @Tabs-spacer-height, @color-white 3px);
		cursor: initial;
		&:hover {
			background-color: @color-white;
			.Tabs-mixin-tab-arrow-background();
		}
		&:active {
			box-shadow: none;
		}
		.Tabs-mixin-tab-line();
	}

	&-Tab-is-active-and-open {
		border-bottom: 1px solid @color-white;
		z-index: 1; // Safari fix
		&:hover {
			.Tabs-mixin-tab-arrow-background();
		}
		.Tabs-mixin-tab-line();
	}

	&-content {
		border-top: 1px solid @color-default-borderColor;
		padding-top: 10px; // TODO: use our default gutter
	}
}

.Tabs-mixin-tab-line(@color-Tab-line: @color-primary) {
	.lucid-Tabs-Tab-arrow {
		.lucid-Tabs-Tab-arrow-tab-line {
			fill: @color-Tab-line;
		}
	}
}

.Tabs-mixin-tab-arrow-background(@color-Tab-background: @color-white) {
	.lucid-Tabs-Tab-arrow {
		.lucid-Tabs-Tab-arrow-background {
			fill: @color-Tab-background;
		}
	}
}
