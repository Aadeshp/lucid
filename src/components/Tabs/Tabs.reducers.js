export default {
	onSelect(state = {}, index) {
		return {
			...state,
			selectedIndex: index
		}
	}
}

