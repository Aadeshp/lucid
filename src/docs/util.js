import _ from 'lodash';
import { markdown } from 'markdown';

export function toMarkdown(md) {
	if (md) {
		return {
			__html: markdown.toHTML(md),
		};
	}

	return {
		__html: '',
	};
}

export function handleHighlightCode() {
	if (window.hljs) {
		//eslint-disable-line
		_.each(document.querySelectorAll('pre code'), block => {
			hljs.highlightBlock(block); //eslint-disable-line
		});
	}
}

export function sanitizeExamplePath(path) {
	return _.flow(
		x => _.split(x, '/'),
		x => _.compact(x),
		x => _.drop(x),
		x => _.join(x, '-'),
		x => _.split(x, '.'),
		x => _.join(x, '-')
	)(path);
}

export function stripIndent(text) {
	if (!text) {
		return '';
	}

	const sigLines = text.split(/\n/).filter(line => !/^\s*$/.test(line));
	const leadingWhitespaces = sigLines.map(line => /^(\s*)/g.exec(line)[1]);
	const minimumLeadingWhitespace = leadingWhitespaces.reduce(
		(min, line) => (line.length < min.length ? line : min),
		leadingWhitespaces[0]
	);
	return text
		.split(/\n/)
		.map(line => line.replace(minimumLeadingWhitespace, ''))
		.join('\n');
}
