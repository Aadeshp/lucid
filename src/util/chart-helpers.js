import _ from 'lodash';
import d3TimeFormat from 'd3-time-format';

/**
 * groupByFields
 *
 * This will return the data in a similar format to d3Shape.stack
 * but without the stacking of the data.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function groupByFields(collection, fields) {
	const fieldsArray = [].concat(fields);

	return _.map(fieldsArray, (field) => {
		return _.map(collection, (item) => item[field]);
	});
}

/**
 * toPairs
 *
 * Takes a collection of data and pairs up some x and y data.
 *
 * @param {object[]} collection
 * @param {string} xField
 * @param {string} yField
 * @return {array} - E.g. [[x, y], [x, y]]
 */
export function toPairs(collection, xField, yField) {
	return _.map(collection, (item) => [item[xField], item[yField]]);
}

/**
 * byFields
 *
 * Takes a collection of data and returns an array of all the fields off that
 * collection.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {array}
 */
export function byFields(collection, fields) {
	const fieldArray = [].concat(fields);

	return _.reduce(fieldArray, (acc, field) => {
		return acc.concat(_.map(collection, field));
	}, []);
}

/**
 * minByFields
 *
 * Returns the minimum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function minByFields(collection, fields) {
	return _.min(byFields(collection, fields));
}

/**
 * maxByFields
 *
 * Returns the maximum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFields(collection, fields) {
	return _.max(byFields(collection, fields));
}

/**
 * maxByFieldsStacked
 *
 * Returns the max sum of a set of fields from a collection
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFieldsStacked(collection, fields) {
	const fieldArray = [].concat(fields);

	const sums = _.reduce(collection, (acc, item) => {
		const sum = _.chain(item)
			.pick(fieldArray)
			.toArray()
			.sum()
			.value();
		return acc.concat(sum);
	}, []);

	return _.max(sums);
}

/**
 * discreteTicks
 *
 * Returns `count` evenly spaced, representative values from the `array`.
 *
 * @param {array} array
 * @param {number} size - should be greater than 1
 * @return {array}
 */
export function discreteTicks(array, count) {
	if (!array || _.isNil(count) || array.length <= count) {
		return array;
	}

	const step = (array.length - 1) / Math.max(1, count - 1);

	return _.reduce(_.times(count), (acc, n) => {
		return acc.concat(array[Math.round(n  * step)]);
	}, []);
}

/**
 * transformFromCenter
 *
 * Scaling paths from their center is tricky. This function
 * helps do that be generating a translate/scale transform
 * string with the correct numbers.
 *
 * @param {number} x - the x data point where you want the path to be centered at
 * @param {number} y - the y data point where you want the path to be centered at
 * @param {number} xCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} yCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} scale - number to scale to, 2 would be 2x bigger
 * @return {string} - transform string
 */
export function transformFromCenter(x, y, xCenter, yCenter, scale) {
	return `translate(${((1 - scale) * xCenter) + (x - xCenter)}, ${((1 - scale) * yCenter) + (y - yCenter)}) scale(${scale})`;
}

const formatMillisecond = d3TimeFormat.timeFormat('.%L');
const formatSecond = d3TimeFormat.timeFormat(':%S');
const formatMinute = d3TimeFormat.timeFormat('%I:%M');
const formatHour = d3TimeFormat.timeFormat('%I %p');
const formatDay = d3TimeFormat.timeFormat('%a %d');
const formatWeek = d3TimeFormat.timeFormat('%b %d');
const formatMonth = d3TimeFormat.timeFormat('%B');
const formatYear = d3TimeFormat.timeFormat('%Y');

export function multiFormat(date) {
	return (d3TimeFormat.timeSecond(date) < date ? formatMillisecond
		: d3TimeFormat.timeMinute(date) < date ? formatSecond
		: d3TimeFormat.timeHour(date) < date ? formatMinute
		: d3TimeFormat.timeDay(date) < date ? formatHour
		: d3TimeFormat.timeMonth(date) < date ? (d3TimeFormat.timeWeek(date) < date ? formatDay : formatWeek)
		: d3TimeFormat.timeYear(date) < date ? formatMonth
		: formatYear)(date);
}

