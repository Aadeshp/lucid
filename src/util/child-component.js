import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {
			isRenderNull: true,

			/**
			 * findInChildren
			 *
			 * Returns all elements with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} elements with this component type
			 */
			findInChildren(elements) {
				let component = this;
				let ownElements = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						ownElements.push(element);
					}
				});
				return ownElements;
			},

			/**
			 * findInChildrenAsProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} `props` objects of elements with this component type
			 */
			findInChildrenAsProps(elements) {
				let component = this;
				return _.map(component.findInChildren(elements), 'props');
			},

			/**
			 * findInProps
			 *
			 * Given a set of parent props this will return only the props for this
			 * component type
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of normalized props
			 */
			findInProps(parentProps) {
				// Extract out the defined `childProps` from the parent class
				const currentProps = _.chain(parentProps)
					.get(definition.propName) // grab the prop we care about, e.g. "Child"
					.thru(x => [x]) // wrap in an array
					.flatten()
					.filter(x => x) // remove falsey values
					.value();

				return _.map(currentProps, (childProp) => {
					if (_.isPlainObject(childProp)) {
						return childProp;
					} else {
						return {
							children: childProp
						};
					}
				});
			},

			/**
			 * findInAllAsProps
			 *
			 * Given a parent props object, this will return a mashup of all the
			 * children and/or props for the current component type. This is
			 * particularly useful for providing a flexible component API. It allows
			 * consumers to either pass stuff through props or children.
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of `props`
			 */
			findInAllAsProps(parentProps) {
				let component = this;

				const propsFromProps = component.findInProps(parentProps);
				const propsFromChildren = component.findInChildrenAsProps(parentProps.children);

				return propsFromProps.concat(propsFromChildren);
			}
		}
	});
}

export function rejectNullElements(children) {
	return _.reject(React.Children.toArray(children), (node) => (React.isValidElement(node) && node.type.isRenderNull));
}
