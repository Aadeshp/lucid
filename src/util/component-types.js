import React from 'react';
import _ from 'lodash';

// creates a React component
export function createClass(definition={}) {
	const {
		_isPrivate = false,
		statics = {},
		components = {},
		reducers = {},
		selectors = {},
		propName = null,
		propTypes = {},
		render = () => null,
		...restDefinition
	} = definition;

	const newDefinition = {
		...restDefinition,
		statics: {
			...statics,
			...components,
			_isPrivate,
			reducers,
			selectors,
			propName,
		},
		propTypes: _.assign({}, propTypes, _.mapValues(definition.components, () => React.PropTypes.any)),
		render,
	};

	newDefinition.statics.definition = newDefinition;

	return React.createClass(newDefinition);
}

// return all elements matching the specified types
export function filterTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.filter(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return all elements not matching the specified types
export function rejectTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.reject(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return an array of elements (of the given type) for each of the values
export function createElements(type, values=[]) {
	return _.reduce([].concat(values), (elements, typeValue) => {
		if (React.isValidElement(typeValue) && typeValue.type === type) {
			return elements.concat(typeValue);
		} else if (_.isPlainObject(typeValue) && !React.isValidElement(typeValue)) {
			return elements.concat(React.createElement(type, typeValue));
		} else if (_.isUndefined(typeValue)) {
			return elements;
		} else {
			return elements.concat(React.createElement(type, null, typeValue));
		}
	}, []);
}

// return all elements found in props and children of the specified types
export function findTypes(props, types=[]) {
	// get elements from props (using type.propName)
	const elementsFromProps = _.reduce(_.castArray(types), (acc, type) => {
		if (!_.isUndefined(type.propName)) {
			const propMatches = _.flatten(_.values(_.pick(props, type.propName)));
			return acc.concat(createElements(type, propMatches));
		}
		return acc;
	}, []);

	// return elements from props and elements from children
	return elementsFromProps.concat(filterTypes(props.children, types));
}

// return the first element found in props and children of the specificed type(s)
export function getFirst(props, types, defaultValue) {
	return _.first(findTypes(props, types)) || defaultValue;
}

// omit props defined in propTypes of the given type
// and any extra keys given in third argument
export function omitProps(props, type, keys = []) {
	return _.omit(props, _.keys(type.propTypes).concat(keys).concat(['initialState']));
}
