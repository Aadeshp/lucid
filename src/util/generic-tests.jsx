import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, getDefaultProps=_.noop) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe(`[common]`, () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f'
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			assert.deepEqual(wrapper.first().prop('style'), style);
		});

		it('should pass through `className`', () => {
			let expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			let expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let parentClasses = wrapper.first().prop('className').split(' ');
			let childrenClasses = wrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			let allClasses = parentClasses.concat(childrenClasses);

			assert(_.every(allClasses, (className) => {
				return _.includes(className, 'lucid-' + Component.displayName);
			}));
		});

		it('should only use onX convention for function proptypes', () => {
			assert(_.every(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				let isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				let isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`
				if (isProbablyFunction && !isAny && !_.startsWith(key, 'on')) {
					return false
				}

				return true;
			}));
		});

		it('should be available as an exported module from index.js', () => {
			assert(lucid[Component.displayName]);
		});
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM(`[icon]`, () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `uniqueId` and `event` prop', () => {
			const expectedUniqueId = 32;
			const props = {
				uniqueId: expectedUniqueId,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { uniqueId, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(uniqueId, expectedUniqueId, 'incorrect or missing uniqueId');
		});
	});
}
