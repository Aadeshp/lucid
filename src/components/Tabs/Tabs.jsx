import _ from 'lodash';
import React from 'react';
import classNames from 'classnames';
import { bindClassNames } from '../../util/style-helpers';
import { createLucidComponentDefinition } from '../../util/component-definition';
import reducers from './Tabs.reducers';

const boundClassNames = bindClassNames('Tabs');

const {
	any,
	string,
	number,
	bool,
	func,
} = React.PropTypes;

/**
 *
 * {"categories": ["navigation"]}
 *
 * `Tabs` provides tabbed navigation. It has a flexible interface that allows
 * tab content to be passed as regular React children or through props.
 */
const Tabs = React.createClass(createLucidComponentDefinition({
	childProps: {
		Tab: {
			isSelected: bool
		},
		Title: null // we're only interested in children
	},

	reducers,

	propTypes: {
		/**
		 * Styles that are passed through to the root container.
		 */
		style: any,

		/**
		 * While you can pass tab content through this prop, it's better to use the
		 * child component instead. See the examples for details.
		 */
		Tab: any,

		/**
		 * Title can be a prop on `Tabs.Tab` or a child of `Tabs.Tab`. See the
		 * examples for details.
		 */
		Title: any,

		/**
		 * Class names that are appended to the defaults.
		 */
		className: string,

		/**
		 * Index of the currently selected tab.
		 */
		selectedIndex: number,

		/**
		 * Callback for when the user clicks a tab. Called with the index of the
		 * tab that was clicked.
		 */
		onSelect: func,

		/**
		 * If `true` then the active tab will appear open on the bottom.
		 */
		hasOpenTabs: bool,
	},

	getDefaultProps() {
		return {
			style: null,
			className: null,
			selectedIndex: 0,
			onSelect: _.noop,
			hasOpenTabs: true,
		};
	},

	render() {
		const {
			className,
			onSelect,
			selectedIndex,
			style,
			hasOpenTabs,
			...passThroughs,
		} = this.props;

		const tabChildProps = Tabs.Tab.findInAllAsProps(this.props);

		const rootClasses = classNames(className, boundClassNames('~'));

		return (
			<div
				{...passThroughs}
				style={style}
				className={rootClasses}
			>
				<ul className={boundClassNames('bar')}>
					{_.map(tabChildProps, (tabChildProp, index) => {
						return (
							<li
								className={boundClassNames('bar-item', {
									'bar-item-is-active': index === selectedIndex,
									'bar-item-is-active-and-open': hasOpenTabs && index === selectedIndex
								})}
								key={index}
								onClick={_.partial(onSelect, index)}
							>
								{_.get(_.first(Tabs.Title.findInAllAsProps(tabChildProp)), 'children', '')}
							</li>
						)
					})}
				</ul>
				<div className={boundClassNames('content')}>
					{tabChildProps[selectedIndex].children}
				</div>
			</div>
		);
	}
}));

export default Tabs;
