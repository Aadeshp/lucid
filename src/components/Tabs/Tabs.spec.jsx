import sinon from 'sinon';
import React from 'react';
import assert from 'assert';
import { shallow } from 'enzyme';
import { common } from '../../util/generic-tests';

import Tabs from './Tabs';

describe('Tabs', () => {
	common(Tabs);

	describe('props', () => {
		it('Tab as children', () => {
			const wrapper = shallow(
				<Tabs>
					<Tabs.Tab>Foo Content</Tabs.Tab>
					<Tabs.Tab>Two</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Foo Content')
		});

		it('Tab as props', () => {
			const wrapper = shallow(
				<Tabs Tab={[{children: 'Bert'}, {children: 'Ernie'}]} />
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab').length, 2);
			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Bert');
		});

		it('Tab as props with Title', () => {
			const wrapper = shallow(
				<Tabs Tab={[
					{ Title: 'Coolest', children: 'Bert' },
					{ Title: 'Not so cool', children: 'Ernie' },
				]} />
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab').length, 2);
			assert.equal(wrapper.find('.lucid-Tabs-Tab-is-active').text(), 'Coolest');
			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Bert');
		});

		it('Title as props', () => {
			const wrapper = shallow(
				<Tabs>
					<Tabs.Tab Title='Froyo'>Yolo fo sho</Tabs.Tab>
					<Tabs.Tab>Broyoyo</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab').first().text(), 'Froyo');
		});

		it('Title as children', () => {
			const wrapper = shallow(
				<Tabs>
					<Tabs.Tab>
						<Tabs.Title>Froyo</Tabs.Title>
						Yolo fo sho
					</Tabs.Tab>
					<Tabs.Tab>Broyoyo</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab').first().text(), 'Froyo');
		});

		it('selectedIndex', () => {
			const wrapper = shallow(
				<Tabs selectedIndex={1}>
					<Tabs.Tab Title='Lollipop'>Yuck</Tabs.Tab>
					<Tabs.Tab Title='Slurpee'>Yum</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab-is-active').text(), 'Slurpee');
			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Yum');
		});

		it('Tab.isSelected', () => {
			const wrapper = shallow(
				<Tabs>
					<Tabs.Tab Title='Lollipop'>Yuck</Tabs.Tab>
					<Tabs.Tab isSelected={true} Title='Slurpee'>Yum</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab-is-active').text(), 'Slurpee');
			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Yum');
		});

		it('last Tab.isSelected beats selectedIndex', () => {
			const wrapper = shallow(
				<Tabs selectedIndex={0}>
					<Tabs.Tab Title='One'>One content</Tabs.Tab>
					<Tabs.Tab isSelected={true} Title='Two'>Two content</Tabs.Tab>
					<Tabs.Tab isSelected={true} Title='Three'>Three content</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab-is-active').text(), 'Three');
			assert.equal(wrapper.find('.lucid-Tabs-content').text(), 'Three content');
		});

		it('onSelect', () => {
			const onSelect = sinon.spy();
			const clickEvent = 'event';
			const wrapper = shallow(
				<Tabs onSelect={onSelect}>
					<Tabs.Tab>One</Tabs.Tab>
					<Tabs.Tab>Two</Tabs.Tab>
				</Tabs>
			);

			wrapper.find('.lucid-Tabs-Tab').at(1).simulate('click', clickEvent);

			assert(onSelect.called);
			assert.equal(onSelect.args[0][0], 1);
			assert.equal(onSelect.args[0][1]['event'], clickEvent);
			assert.deepEqual(onSelect.args[0][1]['props'], {children: 'Two'});
		});

		it('isOpen', () => {
			const wrapper = shallow(
				<Tabs isOpen={false} selectedIndex={0}>
					<Tabs.Tab Title='Lollipop'>Yuck</Tabs.Tab>
					<Tabs.Tab Title='Slurpee'>Yum</Tabs.Tab>
				</Tabs>
			);

			assert.equal(wrapper.find('.lucid-Tabs-Tab-is-active-and-open').length, 0);
		});
	});
});
