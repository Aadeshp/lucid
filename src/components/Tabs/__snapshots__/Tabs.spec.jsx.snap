exports[`Tabs [common] example testing should match snapshot for 1-title-as-prop 1`] = `
<div>
  <Tabs>
    <Tabs.Tab
      Title="One">
      One content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Two"
      isDisabled={true}>
      Two content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Three">
      Three content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Five">
      Four content
    </Tabs.Tab>
  </Tabs>
</div>
`;

exports[`Tabs [common] example testing should match snapshot for 2-title-as-child 1`] = `
<div>
  <Tabs>
    <Tabs.Tab>
      <Tabs.Title>
        One
      </Tabs.Title>
      One content
    </Tabs.Tab>
    <Tabs.Tab>
      <Tabs.Title>
        Two
      </Tabs.Title>
      Two content
    </Tabs.Tab>
    <Tabs.Tab>
      <Tabs.Title>
        Three
      </Tabs.Title>
      Three content
    </Tabs.Tab>
    <Tabs.Tab
      isDisabled={true}>
      <Tabs.Title>
        Four
      </Tabs.Title>
      Four content
    </Tabs.Tab>
  </Tabs>
</div>
`;

exports[`Tabs [common] example testing should match snapshot for 3-progression 1`] = `
<div>
  <Tabs
    isOpen={false}
    isProgressive={true}>
    <Tabs.Tab
      Title="One">
      One content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Two">
      Two content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Three">
      Three content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Disabled"
      isDisabled={true}>
      Disabled Content
    </Tabs.Tab>
    <Tabs.Tab
      Title="Five"
      isDisabled={true}>
      Four content
    </Tabs.Tab>
  </Tabs>
</div>
`;

exports[`Tabs [common] example testing should match snapshot for 4-complex-title-as-child 1`] = `
<div>
  <Tabs
    hasMultilineTitle={true}>
    <Tabs.Tab>
      <Tabs.Title>
        One
        <br />
        <Button
          hasOnlyIcon={false}
          isActive={false}
          isDisabled={false}
          onClick={[Function]}
          type="button">
          Button
        </Button>
      </Tabs.Title>
      One content
    </Tabs.Tab>
    <Tabs.Tab>
      <Tabs.Title>
        Two
        <br />
        Line Two
        <SearchField />
      </Tabs.Title>
      Two content
    </Tabs.Tab>
    <Tabs.Tab>
      <Tabs.Title>
        <RadioGroup
          selectedIndex={1}>
          <RadioButton
            isDisabled={false}
            isSelected={false}
            onSelect={[Function]}>
            <RadioGroup.Label>
              Captain America
            </RadioGroup.Label>
          </RadioButton>
          <RadioButton
            isDisabled={false}
            isSelected={false}
            onSelect={[Function]}>
            <RadioGroup.Label>
              Iron Man
            </RadioGroup.Label>
          </RadioButton>
          <RadioButton
            isDisabled={false}
            isSelected={false}
            onSelect={[Function]}>
            <RadioGroup.Label>
              Thor
            </RadioGroup.Label>
          </RadioButton>
        </RadioGroup>
      </Tabs.Title>
      Three content
    </Tabs.Tab>
    <Tabs.Tab
      isDisabled={true}>
      <Tabs.Title>
        Four
      </Tabs.Title>
      Four content
    </Tabs.Tab>
  </Tabs>
</div>
`;

exports[`Tabs [common] example testing should match snapshot for 5-tab-as-prop 1`] = `
<div>
  <Tabs
    Tab={
      Array [
        Object {
          "Title": "Bert",
          "children": "Bert",
        },
        Object {
          "Title": "Ernie",
          "children": "Ernie",
        },
      ]
    } />
</div>
`;
